query.go

```diff
@@ -778,120 +772,272 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
+		numFilters, stopHash,
+	)
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't get cfheader for block %s "+
-			"from database", blockHash)
+		return nil, fmt.Errorf("unable to get %d blockheader "+
+			"ancestors for stopHash=%d: %v", numFilters,
+			stopHash, err)
+	}
+
+	if len(blockHeaders) != int(numFilters)+1 {
+		return nil, fmt.Errorf("expected %d block headers, got %d",
+			numFilters+1, len(blockHeaders))
+	}
+
+	filterHeaders, _, err := s.RegFilterHeaders.FetchHeaderAncestors(
+		numFilters, stopHash,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get %d filter header "+
+			"ancestors for stopHash=%d: %v", numFilters, stopHash,
+			err)
+	}
+
+	if len(filterHeaders) != int(numFilters)+1 {
+		return nil, fmt.Errorf("expected %d filter headers, got %d",
+			numFilters+1, len(filterHeaders))
+	}
+
+	// We create a height index such that we can easily index into our
+	// header slices for a given block hash in the received response,
+	// without consulting the database. This also keeps track of for which
+	// blocks we are still awaiting a response. We start at index=1, as 0
+	// is for the block startHeight-1, only needed for validation.
+	heightIndex := make(map[chainhash.Hash]int)
+	for i := 1; i < len(blockHeaders); i++ {
+		block := blockHeaders[i]
+		heightIndex[block.BlockHash()] = i
+	}
+
+	// We'll immediately respond to the caller with the requested filter
+	// when it is received, so we make a channel to be notified when it's
+	// ready.
+	filterChan := make(chan *gcs.Filter, 1)
+
+	// We acquire the mutex ensuring we don't have several redundant
+	// CFilter queries running in parallel.
+	s.mtxCFilter.Lock()
+
+	// Since another request might have added the filter to the cache while
+	// we were waiting for the mutex, we do a final lookup before starting
+	// our own query.
+	filter, err = s.getFilterFromCache(&blockHash, dbFilterType)
+	if err == nil && filter != nil {
+		s.mtxCFilter.Unlock()
+		return filter, nil
+	}
+	if err != nil && err != cache.ErrElementNotFound {
+		s.mtxCFilter.Unlock()
+		return nil, err
 	}
 
 	// With all the necessary items retrieved, we'll launch our concurrent
 	// query to the set of connected peers.
-	s.queryPeers(
-		// Send a wire.MsgGetCFilters
-		wire.NewMsgGetCFilters(filterType, height, &blockHash),
+	log.Debugf("Fetching filters for heights=[%v, %v], stophash=%v",
+		startHeight, stopHeight, stopHash)
 
-		// Check responses and if we get one that matches, end the
-		// query early.
-		func(sp *ServerPeer, resp wire.Message, quit chan<- struct{}) {
-			switch response := resp.(type) {
-			// We're only interested in "cfilter" messages.
-			case *wire.MsgCFilter:
-				// Only keep this going if we haven't already
-				// found a filter, or we risk closing an
-				// already closed channel.
-				if filter != nil {
-					return
-				}
+	go func() {
+		defer s.mtxCFilter.Unlock()
+		defer close(filterChan)
+
+		s.queryPeers(
```

Comment:
Given that we may now be requesting a series of headers from our peers, we may want to consider using `queryBatch` here, as it'll spread out the load amongst all of our peers allowing the fastest ones to consume most of the queries.

Reply:
It is actually just a single query (GetCFilters), so not sure if `queryBatch` applies?